contracts/KlerosLiquidProxy.sol

```diff
@@ -0,0 +1,245 @@
+            "Sender cannot mediate"
+        );
+        emit SubmittedEvidenceToKlerosArbitrator(
+            claimIndex,
+            msg.sender,
+            evidence
+        );
+        emit Evidence(klerosArbitrator, claimIndex, msg.sender, evidence);
+    }
+
+    function appealKlerosArbitratorRuling(uint256 claimIndex)
+        external
+        payable
+        override
+        onlyClaimForwardedToKleros(claimIndex)
+    {
+        (, , address claimant, , , , ) = claimsManager.claims(claimIndex);
+        // Ruling options
+        // 0: Kleros refused to arbitrate or ruled that it's not appropriate to
+        // pay out the claim or the settlement. We allow both parties to appeal this.
+        // 1: Pay the claim. Only the mediator can appeal this.
+        // 2: Pay the settlement. Only the claimant can appeal this.
+        // We don't check the dispute status (if it's appealable), as the appeal() call
+        // below should revert in that case anyway.
+        if (msg.sender == claimant) {
+            require(
+                klerosArbitrator.currentRuling(
+                    claimIndexToDisputeId[claimIndex]
+                ) != 1,
+                "Ruling agrees with claimant"
+            );
+        } else if (claimsManager.isManagerOrMediator(msg.sender)) {
+            require(
+                klerosArbitrator.currentRuling(
+                    claimIndexToDisputeId[claimIndex]
+                ) != 2,
+                "Ruling agrees with mediator"
+            );
+        } else {
+            revert("Only parties can appeal");
+        }
+        emit AppealedKlerosArbitratorRuling(
+            claimIndex,
+            msg.sender,
+            claimIndexToDisputeId[claimIndex]
+        );
+        klerosArbitrator.appeal{value: msg.value}(
+            claimIndexToDisputeId[claimIndex],
+            klerosArbitratorExtraData // Unused in KlerosLiquid
+        );
+    }
+
+    function rule(uint256 disputeId, uint256 ruling) external override {
+        uint256 claimIndex = disputeIdToClaimIndex[disputeId];
+        require(claimIndex != 0, "No dispute for sender to rule");
+        emit Ruling(IArbitrator(msg.sender), disputeId, ruling);
+        // should revert if ruling > type(ArbitratorDecision).max
+        IClaimsManager.ArbitratorDecision decision = IClaimsManager
+            .ArbitratorDecision(ruling);
+        claimsManager.resolveDispute(claimIndex, decision);
+    }
+
+    function executeRuling(uint256 claimIndex)
+        external
+        override
+        onlyClaimForwardedToKleros(claimIndex)
+    {
+        IExtendedKlerosArbitrator(address(klerosArbitrator)).executeRuling(
+            claimIndexToDisputeId[claimIndex]
+        );
+    }
+
+    function arbitrationCost() external view override returns (uint256) {
+        return klerosArbitrator.arbitrationCost(klerosArbitratorExtraData);
+    }
+
+    function appealCost(uint256 claimIndex)
```

Comment:
I think we should also add the `appealPeriod()` proxy function. It looks useful for determining the deadline of the appeal action.

```solidity
function appealPeriod(uint _disputeID) public view returns(uint start, uint end) {
  Dispute storage dispute = disputes[_disputeID];
  if (dispute.period == Period.appeal) {
    start = dispute.lastPeriodChange;
    end = dispute.lastPeriodChange + courts[dispute.subcourtID].timesPerPeriod[uint(Period.appeal)];
  } else {
    start = 0;
    end = 0;
  }
}
```

Reply:
At first glance it feels like this logic should be implemented on the frontend but I'll look into it

Reply:
It seems like all the data needed to do this calculation is already exposed so implementing this at the contract level feels a bit too redundant to me

Reply:
Just in case I wasn’t 100% clear, this `appealPeriod()` function already exists in the KlerosLiquid.sol contract.

Alright and just to be on the same page as you w.r.t. the data needed for this calculation, afaict it can be done by calling the `getSubCourt()` function which returns the `timesPerPeriod` (to be used in combination with the dispute `period`). And in order to make that call, we need the `subCourtId` which I guess is a known/hardcoded value? If not it seems the frontend would need to load the `klerosArbitratorExtraData` from the contract and parse it to get the `subCourtId`?

Reply:
> Just in case I wasn’t 100% clear, this appealPeriod() function already exists in the KlerosLiquid.sol contract.

Ah okay, I misunderstood. I thought you were talking about the logic above the be implemented in KlerosLiquidProxy

`klerosArbitratorExtraData` is an immutable variable, which means it's safe to hardcode into the frontend (because it's similar to hardcoding the KlerosLiquidProxy address, which you will be doing). My assumption is that you will probably want to fetch `timesPerPeriod` (with the subcourt ID) anyway to display remaining time for phases other than the appeal.

I'll add the wrapped view function anyway because it doesn't hurt.

Reply:
https://github.com/api3dao/claims-manager/pull/27/commits/c3236f296f67f3d1592d1062407e9b5c5ae764f4